### DISCUSSION TITLE
How to use `call_resumable` together with `consume_fuel`?

### DISCUSSION THREAD

**Discussion by tomaka** [wasmi-labs/wasmi#691] [Q&A]
I'd like to call a Wasm function using the relatively new `call_resumable` feature, while at the same time use the `Config::consume_fuel(true)` feature. When the VM runs out of fuel, I would like the execution to be interrupted so that I can resume it later.

My question is: how do these two features work together?

I can see two options, and I don't know which one to expect.

- When an `OutOfFuel` trap happens, `call_resumable` returns a `ResumableCall::Resumable` where the `host_error` lets me know what kind of trap happened. In that situation, does that mean that every trap leads to `ResumableCall::Resumable`?
- When an `OutOfFuel` trap happens, `call_resumable` returns an error. However, in that situation, I don't know how to resume the execution if I simply want to re-fuel the store and ignore the trap.

Thanks for the help

**Comment by Robbepop** [comment]
As of now resumable calls can only be resumed for errors originating from called host functions. If an `OutOfFuel` trap is raised in Wasm code the trap is not resumable. If however, a called host functions manually consumes fuel via the `Caller` parameter that leads to an `OutOfFuel` trap then the `OutOfFuel` trap is originating from the host and therefore the call is resumable.

**Reply by tomaka** [reply]
Thanks for the answer

If I understand correctly, there's no way to do in wasmi something similar to this wasmtime behavior: https://docs.rs/wasmtime/latest/wasmtime/struct.Store.html#method.out_of_fuel_async_yield?

**Reply by Robbepop** [reply]
Correct. I was not aware of this configuration in Wasmtime. Maybe we can add something similar to `wasmi`. Although there is no notion of "async" calls in `wasmi`.

**Reply by Robbepop** [reply]
Please create an issue for `wasmi` with your constraints and usage scenario if you need this to be implemented.
The more info we get the better design we will find.

**Reply by Lucky4Luuk** [reply]
I'm looking for this exact feature. I know it's only been 4 days, but is there any progress on this? This seems to be the only reasonable choice right now for wasm in a no_std environment

**Reply by tomaka** [reply]
https://github.com/paritytech/wasmi/issues/696
